src/shell/optional.js

```diff
@@ -0,0 +1,116 @@
+var Path = require('../path.js');
+var Errors = require('../errors.js');
+var async = require('../../lib/async.js');
+
+var optional = {};
+module.exports = optional;
+
+/**
+ * Create a file if it does not exist, or update access and
+ * modified times if it does. Valid options include:
+ *
+ *  * updateOnly - whether to create the file if missing (defaults to false)
+ *  * date - use the provided Date value instead of current date/time
+ */
+function touch(path, options, callback) {
+  var sh = this;
+  var fs = sh.fs;
+  if(typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  options = options || {};
+  callback = callback || function(){};
+  path = Path.resolve(sh.pwd(), path);
+
+  function createFile(path) {
+    fs.writeFile(path, '', callback);
+  }
+
+  function updateTimes(path) {
+    var now = Date.now();
+    var atime = options.date || now;
+    var mtime = options.date || now;
+
+    fs.utimes(path, atime, mtime, callback);
+  }
+
+  fs.stat(path, function(error, stats) {
+    if(error) {
+      if(options.updateOnly === true) {
+        callback();
+      } else {
+        createFile(path);
+      }
+    } else {
+      updateTimes(path);
+    }
+  });
+}
+
+/**
+ * Concatenate multiple files into a single String, with each
+ * file separated by a newline. The `files` argument should
+ * be a String (path to single file) or an Array of Strings
+ * (multiple file paths).
+ */
+function cat(files, callback) {
+  var sh = this;
+  var fs = sh.fs;
+  var all = '';
+  callback = callback || function(){};
+
+  if(!files) {
+    callback(new Errors.EINVAL('Missing files argument'));
+    return;
+  }
+
+  files = typeof files === 'string' ? [ files ] : files;
+
+  function append(item, callback) {
+    var filename = Path.resolve(sh.pwd(), item);
+    fs.readFile(filename, 'utf8', function(error, data) {
+      if(error) {
+        callback(error);
+        return;
+      }
+      all += data + '\n';
+      callback();
+    });
+  }
+
+  async.eachSeries(files, append, function(error) {
+    if(error) {
+      callback(error);
+    } else {
+      callback(null, all.replace(/\n$/, ''));
+    }
+  });
+};
+
+/**
+ * Convenience non-binding access methods
+ */
+optional.touch = function(sh, path, options, callback) {
+  touch.call(sh, path, options, callback);
+};
+
+optional.cat = function(sh, files, callback) {
+  cat.call(sh, files, callback);
+};
+
+/**
+ * Access to optional methods for shell
```

Comment:
We should probably chat when you have some time to focus my approach to the problem of how to provide access to optional shell methods. This was intended to allow something like:

```javascript
var filer = require('filer');
var optionals = filer.optionals;
var sh = filer.Shell;

optionals.bindTouch(sh);

module.exports = sh;
```

This example is at best pseudocoded, and my approach hasn't really solved this problem yet. I thought it might be a good starting point for discussion though!